lib/auth/auth_with_roles.go

```diff
@@ -2608,6 +2609,36 @@ func (a *ServerWithRoles) SubmitAccessReview(ctx context.Context, submission typ
 	return a.authServer.submitAccessReview(ctx, submission, &identity)
 }
 
+func (a *ServerWithRoles) canFilterRequestableRolesByResource(ctx context.Context, req types.AccessCapabilitiesRequest) (bool, error) {
+	if len(req.RequestableResourceIDs) == 0 {
+		return false, nil
+	}
+	currentCluster, err := a.GetClusterName()
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	for _, resourceID := range req.RequestableResourceIDs {
+		if resourceID.ClusterName != currentCluster.GetClusterName() {
+			// Requested resource is from another cluster, so we can't know
+			// all of the roles which would grant access to it.
+			return false, nil
+		}
+	}
+
+	resources, err := accessrequest.GetResourcesByResourceIDs(ctx, a, req.RequestableResourceIDs)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	for _, resource := range resources {
+		if err := a.context.CheckAccessToResource(resource, types.VerbRead); err != nil {
+			// User doesn't have read access to one or more resources, so returning
+			// requestable roles that rely on it may leak information.
+			return false, nil
+		}
```

Comment:
I don't think this check really does much? the default implicit rules give read access to all resources, but we don't actually allow reading resources unless the resource labels or label expressions allow access